### PR TITLE
Send events using POST

### PR DESCRIPTION
**Pull request by mattagra**
## Pivotal ticket(s)
https://www.pivotaltracker.com/story/show/177824064

## Context
[WT Event Fixes Spec](https://docs.google.com/document/d/1RJJDpNNHmMhM94-EL1Go0TdYQS9czFoBx07d_Q1UYvk/edit#)
> Missing WT events have been a recurring problem since WT events first launched. Adblockers can affect the efficacy of our pixels (and even prevent them from appearing in fullstory). There are some low-hanging-fruit steps we can take to reduce the impact of adblockers and have a higher integrity events dataset. 

## Changes
- Send events to `/wt/t` with `POST` method using window.fetch()
- Fallback to image pixel tracking if `window.fetch()` fails (plan is to remove in the future)
- Send data in POST request body to avoid "URI Too Long" error:

Screen Shot 2021-04-21 at 2 54 32 PM

## Test Cases
Tested on macallan
- Published an rc version to npm with `@clutter/wt@1.2.0-rc.1`
- Pointed landing branch to rc version: https://github.com/clutter/landing/pull/2003
- Updated wt gem for new wt endpoint: https://github.com/clutter/clutter-platform/pull/11408

See events on macallan:
```
select * from wt_events order by created_at desc limit 30
```

Chrome:
Screen Shot 2021-04-23 at 7 23 30 AM

Firefox:
Screen Shot 2021-04-22 at 11 47 19 PM

Safari:
Screen Shot 2021-04-22 at 11 48 39 PM

## Callouts
wt changes: https://github.com/clutter/wt/pull/62

## Future
- Removing pixel fallback (if we don't see many fallback=true requests after we deploy)
- Use sendBeacon with polyfill? Not sure if we still want this if things work well with fetch/POST
- Use prettier
- (Optional) Update circleci config

### CODE REVIEW

File: src/index.js
Comment (bcharna) [inline src/index.js:158]: afaict, the only time this would return false is if the data limit is surpassed, see the note under **Return Value** [here](https://www.w3.org/TR/beacon/#dom-navigator-sendbeacon). given that, I'm a bit skeptical that we'd ever get false returned but I think we don't know for sure. that being said, can we just hit the new POST endpoint instead (via a standard AJAX call) so we don't have to maintain the `track.gif` endpoint? then, we also wouldn't need `beaconUrl` and `trackerUrl`, we can just have the latter.

additionally, not all browsers will have `navigator.sendBeacon`, so I think it would be wise to add a polyfill to avoid crashing in this case, or checking for the existence of the function and falling back to an AJAX call if it doesn't exist.

File: src/index.js
Comment (bcharna) [inline src/index.js:57]: I like this! will it go into `event.metadata`?

**Review by bcharna — COMMENTED** [review]
had just a couple comments. a couple other call-outs:

1. can we see if this seems to not get blocked by Adblock on folk's local machines? I coincidentally was just helping Michael with a WT thing and it was being blocked by his Adblock, so we could test it on his machine (or likely anyone else's)
2. can we ensure that we have the best view of pageloads -> pageviews possible ahead of launching this? i was working with Clayton on some of this, but I think we should be a bit more confident on those charts before shipping so we are confident we have a good way to track how well this change goes.

**Comment by bcharna** [comment]
also, going to pass along to @nopelluhh to do a pass as well, since this is such critical code.

File: src/index.js
Comment (nopelluhh) [reply src/index.js:57]: i'm not sure that i like this (as is); recording this for every event seems excessive, _especially_ if it's going into an already unstructured metadata column. what's the motivation for tracking this/do we need it on a data level/do we need it for every event?

if we do keep this i'd suggest renaming to `wt_client_version` for accuracy.

File: src/index.js
Comment (nopelluhh) [reply src/index.js:158]: @bcharna i believe there's an example of hitting the data limit (via an overly long query string) in the PR description. 

that said, i think the only reason that's likely to happen is due to the how we batch events together. i'm not sure if there's any reason to do that with the beacon API, but if we _can_ remove that it would be a win in terms of complexity.

re: polyfilling, i think we can safely offload that to consumers of the library rather than make that decision for everybody.

File: src/index.js
Comment (nopelluhh) [inline src/index.js:157]: along the lines of brad's browser support callout, this is a newer API. we already use the `qs` library so i'd suggest using that (or fully migrating to `URLSearchParams` if you have time for a bigger change).

File: src/index.js
Comment (nopelluhh) [inline src/index.js:162]: in line with the above comment on when we'd fall back to this, i'm concerned we don't have visibility into when/how often this fallback is happening. it's liable to happen frequently in areas where we track text input (putting aside how silly some of our tracking is on that front).

File: src/index.js
Comment (bcharna) [reply src/index.js:57]: I don't think we have to worry about storage size, but I hear you. I actually think having app version info would be more useful (so not super opinionated on whether we leave wt version info in or not), but that's outside of the scope here.

File: src/index.js
Comment (bcharna) [reply src/index.js:158]: yeah, except the data limits may be different.

also agree batching may not be necessary, but likely something we can punt on for now to keep change size smaller.

fair point on polyfilling – are you proposing we go with my latter suggestion (check for existence of function and hit AJAX if not present)?

File: src/index.js
Comment (bcharna) [reply src/index.js:162]: yeah, great call-out, was thinking something similar. we could add an extra param in the body (that doesn't need to go into the wt event), something like `fallback: true`, and then see frequency by querying request logs.

File: src/index.js
Comment (nopelluhh) [reply src/index.js:158]: proposing we leave as is + add a note in the readme that `sendBeacon` support is required and users may polyfill if neccesary.

Comment (bcharna) [reply src/index.js:158]: ah, I see. sorry, I sort of combined two pieces of feedback in the original comment. so I think to summarize we're saying:

1. end users of wt (for us, landing, platform and today) will polyfill themselves.
2. we'll make a POST request instead of an image request in the case that `this.context.navigator.sendBeacon` returns `false`

Comment (mattagra) [reply src/index.js:158]: Do we want to completely get rid of the image request (i.e. try sendBeacon then try POST)? or try all three (i.e. try sendBeacon then try POST then try image)?

File: src/index.js
Comment (mattagra) [reply src/index.js:57]: yeah, this will be saved in `event.metadata`. this was to help me QA that the new events were saving properly on macallan so I don't feel too strongly about keeping it. i feel ok with removing it because it doesn't add much value outside of the QA use case.

File: src/index.js
Comment (bcharna) [reply src/index.js:158]: I don't see a reason to keep the image request around – all browsers should support POST via standard AJAX. but to clarify, @nopelluhh was proposing only to fallback to the standard AJAX if `sendBeacon` returns false, thus we assume that the function is present (this means we'll want to add a polyfill in our apps that use wt).

File: src/index.js
Comment (mattagra) [reply src/index.js:157]: What's our cutoff for browser support for both `sendBeacon` and `URLSearchParams`?  Their browser compatibility is pretty similar with full support on latest versions except no support on IE. For `sendBeacon`'s data parameter, I'm taking advantage of the `URLSearchParams` object so they kind of go hand-in-hand so I don't think I could use `qs` with `sendBeacon`. I also tried the [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob) object with `sendBeacon` to send the data as JSON but that led to CORS issues. 

i selfishly like the idea that noah suggested to leave as is, add note in the readme that `sendBeacon` and `URLSearchParams` are required and let users polyfill if necessary. 

this also leads back to the fallback question: if we're requiring sendBeacon support, should we just get rid of all fallbacks (POST and image)? i left the image code in there to keep the change small

File: src/index.js
Comment (mattagra) [reply src/index.js:158]: What are your thoughts on a [sendbeacon-polyfill](https://www.npmjs.com/package/sendbeacon-polyfill)? I see from the [source code ](https://github.com/kronicker/sendbeacon-polyfill/blob/master/sendbeacon.js) that it already handles sending as AJAX

Comment (bcharna) [reply src/index.js:158]: it looks like that's a fork of https://github.com/miguelmota/Navigator.sendBeacon, can we see what the difference is between those? it looks like the original repo has more stars/has been updated more recently but have never worked w/ either.

and yep, a lot of cleanup is possible here!

**Comment by mattagra** [comment]
I made the following changes, repushed to Macallan, and re-tested on Chrome, Safari, and Firefox.

Changes:
- Removed adding `wt_version` to all events since it doesn't add much value
- Changed `sendBeacon` to polyfilled `window.fetch()` for increased browser compatibility
- Added `?fallback=true` param to the image pixel fallback so we can measure POST errors

Going to fast-follow these changes so we can unblock marketing:
- Removing pixel fallback (if we don't see many `fallback=true` requests after we deploy)
- Use `sendBeacon` with polyfill? Not sure if we still want this if things work well with fetch/POST
- Use prettier
- (Optional) Update circleci config

> can we see if this seems to not get blocked by Adblock on folk's local machines? I coincidentally was just helping Michael with a WT thing and it was being blocked by his Adblock, so we could test it on his machine (or likely anyone else's)

I got a few people on the team to QA but didn't explicitly call out using AdBlock. I did verify that the fix was not blocked by uBlock on my own machine. I'll get a few more people to test with ad blockers today.

> can we ensure that we have the best view of pageloads -> pageviews possible ahead of launching this? i was working with Clayton on some of this, but I think we should be a bit more confident on those charts before shipping so we are confident we have a good way to track how well this change goes.

Yup, Clayton sent over this [dashboard](https://app.periscopedata.com/app/clutter/841538/Moving-Sandbox?widget=11561400&udv=0). Hoping that we see a noticeable jump up after the release.

**Comment by mattagra** [comment]
Update: Got Deergha to confirm that events are coming through, even with [Adblock](https://getadblock.com/)

**Comment by bcharna** [comment]
great on the QA front, also yeah that's the dashboard we were looking at. I don't think it's perfect, but should provide us at least directional insights.

> Use sendBeacon with polyfill? Not sure if we still want this if things work well with fetch/POST

I still think we should do this as Beacon provides more of a guarantee that requests will go through (if browser window is closed). let's also consider removing batching in the next go.

File: src/index.js
Comment (bcharna) [inline src/index.js:152]: can we switch this to json? suppose can wait until the follow-up release

**Review by bcharna — APPROVED** [review]

**Comment by bcharna** [comment]
also @mattagra, can you update the PR description now that stuff has changed?

File: src/index.js
Comment (mattagra) [reply src/index.js:152]: Yeah good question. I initially thought json wouldn't work because I was getting a CORS error, but the change in platform seems to have addressed it now. I can switch to json in the follow-up.

**Comment by mattagra** [comment]
> can you update the PR description now that stuff has changed?

Done. Also added fast-follows to the PR description.

**Comment by mattagra** [comment]
Interesting, we're seeing a few events fallback to the pixel image with `?fallback=true`: https://app.logz.io/#/goto/8260e1c0ae2376cf9f37cea32bdb3eb2?switchToAccountId=54176. 

Not seeing non-2xx on the /wt/t endpoint though so it might be a problem receiving the response from the server. Also makes me wonder if this would double record events. Seems like `sendBeacon` would fix this since it's not relying on the response to be returned.

**Pull request merged**
